srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/ItemRestoration.php

```diff
@@ -0,0 +1,274 @@
+        $product = $variation_id ? \wc_get_product($variation_id) : \wc_get_product($product_id);
+
+        if ($product->get_sku() != self::ITEM_RESTORATION_SKU) {
+            return $cart_item_data;
+        }
+
+        if (isset($_REQUEST['acore_char_sel']) && isset($_REQUEST['acore_restore_item_sel'])) {
+            $cart_item_data['acore_char_sel'] = $_REQUEST['acore_char_sel'];
+            $cart_item_data['acore_restore_item_sel'] = $_REQUEST['acore_restore_item_sel'];
+            $cart_item_data['acore_item_sku'] = $product->get_sku();
+            /* below statement make sure every add to cart action as unique line item */
+            $cart_item_data['unique_key'] = md5(microtime() . rand());
+        }
+
+        return $cart_item_data;
+    }
+
+    // Render on checkout
+    public static function get_item_data($cart_data, $cart_item = null) {
+        $custom_items = array();
+        if (!empty($cart_data)) {
+            $custom_items = $cart_data;
+        }
+
+        if (!in_array($cart_item['acore_item_sku'], self::$skuList)) {
+            return $custom_items;
+        }
+
+        if (isset($cart_item['acore_char_sel'])) {
+            $ACoreSrv = ACoreServices::I();
+            $charRepo = $ACoreSrv->getCharactersRepo();
+
+            $charId = $cart_item['acore_char_sel'];
+            $itemId = $cart_item['acore_restore_item_sel'];
+
+            $char = $charRepo->findOneByGuid($charId);
+
+            $charName = $char ? $char->getName() : "The character <$charId> does not exist!";
+
+            $custom_items[] = array("name" => 'Character', "value" => $charName);
+            $custom_items[] = array("name" => 'ItemId', "value" => $itemId);
+        }
+
+        return $custom_items;
+    }
+
+    // ADD DATA TO FINAL ORDER META
+    // This is a piece of code that will add your custom field with order meta.
+    public static function add_order_item_meta($item_id, $values, $cart_item_key) {
+        if ($values['acore_item_sku'] != self::ITEM_RESTORATION_SKU) {
+            return;
+        }
+
+        if (isset($values['acore_char_sel']) && isset($values["acore_item_sku"])) {
+            \wc_add_order_item_meta($item_id, "acore_char_sel", $values['acore_char_sel']);
+            \wc_add_order_item_meta($item_id, "acore_restore_item_sel", $values['acore_restore_item_sel']);
+            \wc_add_order_item_meta($item_id, "acore_item_sku", $values['acore_item_sku']);
+        }
+    }
+
+    // CHECK BEFORE PAYMENT
+    public static function checkout_order_processed($order_id, $posted_data) {
+        $logs = new \WC_Logger();
+
+        $order = new \WC_Order($order_id);
+        $items = $order->get_items();
+
+        $soap = ACoreServices::I()->getServerSoap();
+
+        foreach ($items as $item) {
+            if ($item["acore_item_sku"]) {
+                if ($item["acore_item_sku"] == self::ITEM_RESTORATION_SKU) {
+                    $res = $soap->serverInfo();
```

Comment:
I guess most of the time there will only be one `ITEM_RESTORATION_SKU` in the collection, but if there are usually more, it's not necessary to loop and check the server is up for each one, right? Ie. one check for all items seems sufficient.

Also pardon my ignorance - is calling $soap requests a blocking call?

Reply:
Nice question, I think so, indeed later we check the $res result. So, it is not async.

Reply:
I am moving the serverinfo outside the for

Reply:
Looking at the code, may it can be improved, but about checking server info, if you see, there is the return, so we already execute only once the soap check, I will leave as it is.

We could remove the for() but may in another PR where we do only refactoring and when we refactor also the other services.

Thanks for your suggestions ;-)